### PR TITLE
fix(hydrated_bloc): traversals, e2e, list fix

### LINKED ISSUE #1452: Hydrated: type '_InternalLinkedHashMap<dynamic, dynamic>' is not a subtype of type 'Map<String, dynamic>'
**From our Discord**
I am trying to migrate all of my code to 5.0.0, but I am currently running into the following error for HydratedBloc `type '_InternalLinkedHashMap<dynamic, dynamic>' is not a subtype of type 'Map<String, dynamic>'` , this is happening in the `fromJson` method...

**To Reproduce**
Author of issue (me) thinks you just need two nested `Map<String, whatever>` to reproduce this issue.
But some sample code will be awesome to have, and I will write it a bit :wink:  

**Expected behavior**
Everything goes butter smooth

**Logs**
Stacktrace says it happened here:
```dart
@override
State get state {
  if (_state != null) return _state;
  try {
    final stateJson = storage.read(storageToken);
    if (stateJson == null) return super.state;
    return fromJson(Map<String, dynamic>.from(stateJson));
  } on dynamic catch (error, stackTrace) {
    onError(error, stackTrace);
    return super.state;
  }
}
```

### PR DESCRIPTION
**Pull request by orsenkucher**
## Status

**READY**

## Breaking Changes

YES 

## Description

closes #1452 and #1453

## Type of Change

- [ ] ✨ New feature (non-breaking change which adds functionality)
- [x] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: packages/hydrated_bloc/lib/src/hydrated_cubit.dart [hunk 1]
```diff
@@ -101,9 +101,13 @@ mixin HydratedMixin<State> on Cubit<State> {
...
-    if (stateJson != null) {
-      storage.write(storageToken, stateJson).then((_) {}, onError: onError);
+    try {
+      final stateJson = _toJson(state);
+      if (stateJson != null) {
+        storage.write(storageToken, stateJson).then((_) {}, onError: onError);
+      }
+    } on dynamic catch (error, stackTrace) {
+      onError(error, stackTrace);
     }
```
Comment (orsenkucher) [inline packages/hydrated_bloc/lib/src/hydrated_cubit.dart:111]: we have try catch here

File: packages/hydrated_bloc/lib/src/hydrated_cubit.dart [hunk 2]
```diff
@@ -116,25 +120,118 @@ mixin HydratedMixin<State> on Cubit<State> {
...
+  }
+
+  T _cast<T>(dynamic x) => x is T ? x : null;
+
   @override
   void onChange(Change<State> change) {
     if (storage == null) throw const StorageNotFound();
     final state = change.nextState;
-    final stateJson = toJson(state);
+    final stateJson = _toJson(state);
```
Comment (orsenkucher) [inline packages/hydrated_bloc/lib/src/hydrated_cubit.dart:154]: should it be also here?

Comment (felangel) [reply packages/hydrated_bloc/lib/src/hydrated_cubit.dart:154]: yeah I think this should also be wrapped in a try/catch

File: packages/hydrated_bloc/lib/src/hydrated_cubit.dart [hunk 3]
```diff
@@ -163,3 +260,51 @@ mixin HydratedMixin<State> on Cubit<State> {
...
+
+  /// The object that could not be serialized.
+  final Object unsupportedObject;
+
+  /// The exception thrown when trying to convert the object.
+  final Object cause;
+
+  @override
+  String toString() {
+    var safeString = Error.safeToString(unsupportedObject);
```
Comment (felangel) [inline packages/hydrated_bloc/lib/src/hydrated_cubit.dart:297]:
```suggestion
    final safeString = Error.safeToString(unsupportedObject);
```

File: packages/hydrated_bloc/lib/src/hydrated_cubit.dart [hunk 4]
```diff
@@ -163,3 +260,51 @@ mixin HydratedMixin<State> on Cubit<State> {
...
+    String prefix;
+    if (cause != null) {
+      prefix = 'Converting object to an encodable object failed:';
+    } else {
+      prefix = 'Converting object did not return an encodable object:';
+    }
```
Comment (felangel) [inline packages/hydrated_bloc/lib/src/hydrated_cubit.dart:303]:
```suggestion
    final prefix = cause != null
      ? 'Converting object to an encodable object failed:'
      : 'Converting object did not return an encodable object:';
```

File: packages/hydrated_bloc/test/cubits/bad_cubit.dart [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import 'package:hydrated_bloc/hydrated_bloc.dart';
+
+class BadCubit extends HydratedCubit<BadState> {
```
Comment (felangel) [inline packages/hydrated_bloc/test/cubits/bad_cubit.dart:3]: 🤣

**Pull request merged**